WordPress/openverse-api
# Add option to sort search results by `created_on`

# Updated description by @dhruvkb

## Description

This PR adds two query parameters `unstable__sort_by` and `unstable__sort_dir` to sort media items on the basis of the `created_on` field.

- `unstable__sort_by` can take `relevance` (default) or `indexed_on` (sort by `created_on` field)
- `unstable__sort_dir` can take `desc` (newest first, default) or `asc` (oldest first).

## Testing instructions

0. Check out the PR and bring up the services: `just up`.
1. Search for media without sorting, `https://localhost:50280/v1/images/` and note that the results have no specific sort order.
2. Search for media with sorting, `https://localhost:50280/v1/images/?unstable__sort_by=indexed_on` and note that the results still have no specific sort order.
3. Authenticate yourself.
2. Search for media with sorting, `https://localhost:50280/v1/images/?unstable__sort_by=indexed_on` and note that the results are now sorted by `created_on`.
3. See some result on the first page and then some on the last to get a feel for the sorting order and direction. You can also narrow down the query like [Staci did](https://github.com/WordPress/openverse-api/pull/916#pullrequestreview-1296693111).
4. Try the above steps with `unstable__sort_dir=asc` as well.

Also see [update](https://github.com/WordPress/openverse-api/pull/916#issuecomment-1425590655) below.

---

# Original description by @zackkrida

This draft PR should **not be merged.** It's a simple proof of concept to showcase:

- Adding `created_on` to the API responses
- Sorting all searches by `created_on`

It's meant to show how ElasticSearch's built-in sorting functionality works. I would love to see someone expand on this (my python abilities are really limited 😅) in the following ways:

- Make this sorting conditional on a `sort_by=new` query param.
- Optional: Add a `sort_dir=asc/desc` option to configure the direction of sort
- Make sure these query params are undocumented
- Make this feature only accessible to authenticated users
  - Bonus: make access to this feature accessible only by a special user permission with a checkbox in the Django admin
- Write tests

Once those items are completed we could test this against production data! We should also research the integrity of the `created_on` field, and make sure it is a true representation of when a creative work is added to the Openverse catalog and that we never accidentally rewrite that value in production.
